### PR TITLE
[Giga-Net:] Add a priority queue for urgent message processing

### PR DESCRIPTION
**Pull request by ptschip**
This is an improvement from the current method of rotating any priority messages to the front of the queue..for one it won't cause messages to be out of the order that they were received, like rotate will do sometimes. Furthermore, messages will be processed right away, whereas with rotate you will only get priority processing "when" your message queue gets processed next which could still be a while depending on how many other queues need to be processed in front of you and what their queue depths will be.

So far just have the receive queue working...

### CODE REVIEW

**Comment by ptschip** [comment]
Added the priority send queue. Not entirely happy with it, although it does break right through any congestion in the message queues, the only remaining issue with the priority sendqueue is we can still get out of order messages (as is currently the case). The receive queue works fine and solves the potential out of order issue but maybe need a different approach here with the sendqueue...the main difficulty is dealing with the locking orders.  Will leave this in WIP and maybe we should test this out on the giga-net first anyway since there are quite a few changes in a sensitive part of the code.

**Comment by sickpig** [comment]
@ptschip 

it seems to be a formatting problem in `net_processing.cpp` 

```diff
+++ net_processing.cpp_formatted
@@ -2038,9 +2038,9 @@
                 // swap an incomplete message in the following step.
                 if (!pfrom->vRecvMsg.front().complete())
                 {
-                     // LogPrintf("%s: partial message %d of size %d. Recvd bytes: %d\n", pfrom->GetLogName(),
-                     // msgOnQ.nDataPos, msgOnQ.size(), pfrom->currentRecvMsgSize.value);
-                     break;
+                    // LogPrintf("%s: partial message %d of size %d. Recvd bytes: %d\n", pfrom->GetLogName(),
+                    // msgOnQ.nDataPos, msgOnQ.size(), pfrom->currentRecvMsgSize.value);
+                    break;
                 }
                 else
                 {
```

**Comment by ptschip** [comment]
thx, still in WIP...a bit of work left to do

**Comment by sickpig** [comment]
@ptschip `sendeheaders.py` is failing

File: src/net.cpp [hunk 1]
```diff
@@ -1374,6 +1402,69 @@ void ThreadSocketHandler()
                 continue;
             if (FD_ISSET(hSocket, &fdsetSend))
             {
+                // Send priority messages if there any regardless of which peer, taking care to maintain
+                // locking orders.
+                //
+                // Only send one message, the first message, in the send queue by setting hte fSendOne flag, since
```
Comment (sickpig) [inline src/net.cpp]: typo s/hte/the/

Comment (ptschip) [reply src/net.cpp]: done

**Comment by ptschip** [comment]
This PR is working now but after going through the coding exercise I think the priority send queue needs a better solution.  This PR can be used for testing but I'll go back to my original idea and see if I can get that to work instead.

EDIT: Actually thinking about it, keeping the the noderef with the data is going to create a different set of problems...since we still need to send the first two messages now we'll have to instead send the first message from two different queues...so no real savings there...maybe it's just 6 of one half dozen of the other at this point, plus I'll need to introdoce another set of counters and another almost duplicate function for socketsenddata() so why bother...anyway, just thinking outloud.

**Comment by ptschip** [comment]
Ok I think this is pretty done, just to add tests for the iterator swapping.

**Comment by ptschip** [comment]
added tests

File: src/test/prioritymessage_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2014-2015 The Bitcoin Core developers
+// Copyright (c) 2015-2018 The Bitcoin Unlimited developers
```
Comment (sickpig) [inline src/test/prioritymessage_tests.cpp]: copyright 2019

Comment (ptschip) [reply src/test/prioritymessage_tests.cpp]: fixed

**Comment by sickpig** [comment]
@ptschip 

failing due to bug fixed by #1836, I'm going to merge the travis fix so that ideally we could understand what is failing in the unit tests

```
Still running (1 of 50): DOCKER_EXEC LD_LIBRARY_PATH=/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/depends/x86_64-unknown-linux-gnu/lib make -j2 check VERBOSE=1
Still running (2 of 50): DOCKER_EXEC LD_LIBRARY_PATH=/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/depends/x86_64-unknown-linux-gnu/lib make -j2 check VERBOSE=1
```

**Comment by sickpig** [comment]
@ptschip #1836 got merged, if you rebase and push we probably have a chance to see if we are experiencing an actual bug or just a travis hiccup

**Comment by ptschip** [comment]
rebased and squashed

**Comment by sickpig** [comment]
@ptschip 

compiler error in `requestmanager_tests`

```
 CXX      test/test_test_bitcoin-reverselock_tests.o
../../src/test/requestmanager_tests.cpp:57:15: error: no matching member function for call to 'insert'
    ssCommand.insert(ssCommand.begin(), &data[4], &data[16]);
    ~~~~~~~~~~^~~~~~
```

File: src/net.cpp [hunk 3]
```diff
@@ -651,26 +698,37 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
...
+                    // As a safeguard on some peer trying to dominate our networking
+                    // don't allow unlimited simultaneous priority messages. There should in general only be
+                    // one priority message at any one time with the exception of block HEADERS announcements
+                    // which can come relatively close together in time.
+                    if (vPriorityRecvQ.size() <= 5)
+                    {
+                        // Move the this message to the priority queue.
+                        CNetMessage priority;
+                        std::swap(priority, msg);
+                        vRecvMsg.pop_back();
```
Comment (gandrewstone) [inline src/net.cpp]: It was always a little weird to have an incomplete message hanging off the end of the vRecvMsg queue since that message cannot be processed.  Now that we place the message on one of two queues upon completion its even weirder because we need to pull it off of vRecvMsg if its a priority message.  I think that we should clean this up.

It would be just a few LOC to have a member variable named something like "msgInProgress" that holds the incomplete message, and we would add it to vPriorityRecvQ and assign a new empty one (in a manner similar to line 619) only when its complete as part of this if statement, or to vRecvMsg in an "else" path that's taken if either of these nested ifs fails.

File: src/globals.cpp [hunk 4]
```diff
@@ -217,6 +217,14 @@ CNodeSignals g_signals;
 CAddrMan addrman;
 CDoSManager dosMan;
 
+// A message queue used for priority messages such as graheneblocks or other thintype block messages
+std::atomic<bool> fPriorityRecvMsg{false};
+std::atomic<bool> fPrioritySendMsg{false};
+CCriticalSection cs_PriorityRecvQ;
+CCriticalSection cs_PrioritySendQ;
```
Comment (gandrewstone) [inline src/globals.cpp]: Sorry for the nit, but we name all critical sections except for cs_KeyStore (dammit! -- its inherited from core) with an initial lower case: "cs_priorityRecvQ" and "cs_prioritySendQ".  Consistent naming standards saves later developers time by avoiding "undefined variable".

File: src/net.cpp [hunk 5]
```diff
@@ -1398,6 +1461,75 @@ void ThreadSocketHandler()
...
+                            if (!pfrom->vSendMsg.empty())
+                            {
+                                bool fSendTwo = true;
+                                progress += SocketSendData(pfrom, fSendTwo);
+                            }
+                            else
+                                fEmpty = true;
+                        }
+
+                        if (!noderef.get()->fDisconnect && !lock_sendtwo)
```
Comment (gandrewstone) [inline src/net.cpp]: noderef.get() could be replaced with pfrom

File: src/net.cpp [hunk 6]
```diff
@@ -3131,6 +3263,58 @@ CNode::~CNode()
...
+        for (auto &noderef : vPrioritySendQ)
+        {
+            if (noderef.get() == pnode)
+                nItemsPriority++;
+        }
+    }
+    else
+        nItemsPriority = 1;
+
+    // Move the range of priority msg's forward and place the new priority message (pri3) at the
```
Comment (gandrewstone) [inline src/net.cpp]: I am worried about this complexity.  It seems to me like we are cramming 3 separate things onto one list.  I think you are doing this to minimize changes but I think that we need more changes here to keep things clean.  So instead let me propose that we have:
CSerializeData msgSendingInProgress;
std::deque<CSerializeData> vSendMsg;
std::deque<CSerializeData> vPrioritySendMsg;

SocketSendData does:
```
if (msgSendingInProgress size()==0)
{
  if (vPrioritySendMsg.size() > 0)
  {
    msgSendInProgress = std::move(vPrioritySendMsg.front());
    vPrioritySendMsg.pop_front();
  }
  else if (vSendMsg.size() > 0)
  {
    msgSendInProgress = std::move(vSendMsg.front());
    vSendMsg.pop_front();
  }
}
```

Please double check my std::move semantics -- I'm not a big fan of move.  

Note also that this code has a nSendOffset variable that is the currently unsent offset into vSendMsg[0].  Its awkward to have a single variable that special cases one of the elements in a list -- if someone popped the list and forgot to update this variable we would goof the next send.  So with some relatively small changes your PR can clean up some of this ugly code rather than keep it going.

File: src/net.cpp [hunk 7]
```diff
@@ -3196,30 +3380,32 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
...
-        }
-        else
-            it = vSendMsg.insert(vSendMsg.end(), CSerializeData());
+
+    // If the message is a priority message then move to the front of the deque and add
+    // a CNodeRef entry into the vPrioritySendQ. Now that that message is at the front of the queue
+    // all we have to do is emtpy this first item from the vSendMsg queue when we SendMessages().
+    if (IsPriorityMsg(strCommand))
+    {
+        it = vSendMsg.insert(vSendMsg.begin(), CSerializeData());
```
Comment (gandrewstone) [inline src/net.cpp]: with my suggested breakdown this code becomes a really clean if else with 1 line each.  

Just push to the end of either the priority or regular queue depending on IsPriorityMsg()

File: src/net.cpp [hunk 8]
```diff
@@ -3196,30 +3380,32 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
...
-
-    ssSend.GetAndClear(*it);
-    nSendSize.fetch_add((*it).size());
+        ssSend.GetAndClear(*it);
+        nSendSize.fetch_add((*it).size());
+    }
 
-    // If write queue empty, attempt "optimistic write"
-    if (it == vSendMsg.begin())
-        SocketSendData(this);
```
Comment (gandrewstone) [inline src/net.cpp:3231]: I added this code as part of gigaperf because it made a big difference in minimizing message latency.  Perhaps we've fixed whatever was causing the latency though... maybe comment out with a note and we can try it on and off in gigaperf test 2

File: src/protocol.cpp [hunk 9]
```diff
@@ -59,7 +59,7 @@ const char *GET_MEMPOOLSYNCTX = "getmemsynctx";
 // Mempool sync - end section
 const char *XPEDITEDREQUEST = "req_xpedited";
 const char *XPEDITEDBLK = "Xb";
-const char *XPEDITEDTxn = "Xt";
+const char *XPEDITEDTXN = "Xt";
```
Comment (gandrewstone) [inline src/protocol.cpp:62]: thanks!

File: src/streams.h [hunk 10]
```diff
@@ -142,7 +142,7 @@ class CDataStream
             vch.insert(it, first, last);
     }
 
-    void insert(iterator it, const char *first, const char *last)
+    void insert(iterator it, std::string::iterator first, std::string::iterator last)
     {
         if (last == first)
             return;
```
Comment (gandrewstone) [inline src/streams.h:148]: since you are in here how about changing the assert below to DbgAssert(last - first > 0, return).  So in release mode, nothing is inserted if the range you offer is inverted.

**Review by gandrewstone — CHANGES_REQUESTED** [review]

File: src/streams.h (hunk 10 above)
Comment (ptschip) [reply src/streams.h:148]: If we passed a negative value to memcopy won't we potentially corrupt the memory and cause a crash?  As for myself I don't often compile in debug mode because I work in a VM so I wouldn't even know I had a problem until much later, so I'd like to keep them as is if that's ok.

Comment (gandrewstone) [reply src/streams.h:148]: well the DbgAssert would assert in debug mode and return in release mode so the memcpy would never be hit.  but its fine we can keep the assert...

File: src/net.cpp (hunk 8 above)
Comment (ptschip) [reply src/net.cpp:3231]: I think  you might be right...I'll revert for now and we can retest again later.

File: src/globals.cpp (hunk 4 above)
Comment (ptschip) [reply src/globals.cpp]: There are other instances csTxInQ, csBestBlock etc...but they don't use the underscore...I actually prefer all lowercase myself for cs's :) ...maybe we can discuss one day in the dev meeting, but np, will change it for now...

File: src/net.cpp (hunk 5 above)
Comment (ptschip) [reply src/net.cpp]: thanks, done.

File: src/protocol.cpp (hunk 9 above)
Comment (ptschip) [reply src/protocol.cpp:62]: np

File: src/net.cpp (hunk 3 above)
Comment (ptschip) [reply src/net.cpp]: Not sure I understand the problem your outlining... The msg is complete when we swap and move it to the priority queue...it has no extra data dangling off the end,  unless I'm mistaken msg.readData ensures we end up with only the full message and nothing more before going to through the next iteration of reading in the next message.

File: src/net.cpp (hunk 6 above)
Comment (ptschip) [reply src/net.cpp]: I know what you're saying....this is the first approach I took...It seems so simple, have two separate queues, just like the priority recv messages...but it's much more complicated when it comes time to actually send the data in socketsenddata. It's been 6 months since I've looked at it but from what I remember, it would have needed some bigger and more difficult changes to socket send data...and I can't remember why so I'll have to take a closer look again.   Maybe we can get together next week sometime and discuss on Keybase in the meantime I'll refresh my memory.

Comment (ptschip) [reply src/net.cpp]: I think I might have another approach...you're right, the current set of code works, it is ugly and hard to follow...I'm going to just create a different set of counters, really is just sendoffset we need to track, but I"ll need a new socketsenddata() , will call it prioritysocketsenddata()...that's where I was getting stuck before , trying to wed the two together. Will have take a couple of days to work it out and see where it goes.

**Comment by gandrewstone** [comment]
@ptschip formatting...
